### ISSUE TITLE
ranges conflicts with X11 on linux

### ISSUE THREAD

**Issue by SSE4** [ericniebler/range-v3#500]
it seems like impossible to use ranges-v3 together with X11 (or EGL) on Linux (I've reproduced on Ubuntu 16.04, but same issue I've seen on Arch linux)

for example given the following simple test:
```
#include <EGL/egl.h>

#include <iostream>
#include <cstdlib>
#include <vector>
#include <range/v3/all.hpp>

int main()
{
        std::vector<int> test{1, 3, 2};

        auto sorted = std::move(test) | ranges::action::sort;

        return EXIT_SUCCESS;
}
```

throws lots of errors via clang 3.8:
```
#define Bool int
             ^
In file included from main.cpp:7:
In file included from /home/sse4/range-v3/include/range/v3/all.hpp:17:
In file included from /home/sse4/range-v3/include/range/v3/core.hpp:17:
In file included from /home/sse4/range-v3/include/range/v3/begin_end.hpp:21:
In file included from /home/sse4/range-v3/include/range/v3/range_fwd.hpp:20:
/home/sse4/range-v3/include/meta/meta.hpp:1150:32: error: expected ',' or '>' in template-parameter-list
/usr/include/X11/Xlib.h:82:14: note: expanded from macro 'Bool'
#define Bool int
             ^
In file included from main.cpp:7:
In file included from /home/sse4/range-v3/include/range/v3/all.hpp:17:
In file included from /home/sse4/range-v3/include/range/v3/core.hpp:17:
In file included from /home/sse4/range-v3/include/range/v3/begin_end.hpp:21:
In file included from /home/sse4/range-v3/include/range/v3/range_fwd.hpp:20:
/home/sse4/range-v3/include/meta/meta.hpp:1150:36: error: expected unqualified-id
            template <typename Bool, typename... Bools>
                                   ^
/home/sse4/range-v3/include/meta/meta.hpp:1159:24: error: cannot combine with previous 'bool' declaration specifier
        template <bool Bool>
                       ^
/usr/include/X11/Xlib.h:82:14: note: expanded from macro 'Bool'
#define Bool int
             ^
In file included from main.cpp:7:
In file included from /home/sse4/range-v3/include/range/v3/all.hpp:17:
In file included from /home/sse4/range-v3/include/range/v3/core.hpp:17:
In file included from /home/sse4/range-v3/include/range/v3/begin_end.hpp:21:
In file included from /home/sse4/range-v3/include/range/v3/range_fwd.hpp:20:
/home/sse4/range-v3/include/meta/meta.hpp:1160:34: error: expected '(' for function-style cast or type construction
        using not_c = bool_<!Bool>;
                             ~~~~^
/home/sse4/range-v3/include/meta/meta.hpp:1164:28: error: expected a qualified name after 'typename'
        template <typename Bool>
                           ^
/usr/include/X11/Xlib.h:82:14: note: expanded from macro 'Bool'
#define Bool int
             ^
In file included from main.cpp:7:
In file included from /home/sse4/range-v3/include/range/v3/all.hpp:17:
In file included from /home/sse4/range-v3/include/range/v3/core.hpp:17:
In file included from /home/sse4/range-v3/include/range/v3/begin_end.hpp:21:
In file included from /home/sse4/range-v3/include/range/v3/range_fwd.hpp:20:
/home/sse4/range-v3/include/meta/meta.hpp:1164:28: error: expected ',' or '>' in template-parameter-list
/usr/include/X11/Xlib.h:82:14: note: expanded from macro 'Bool'
#define Bool int
             ^
In file included from main.cpp:7:
In file included from /home/sse4/range-v3/include/range/v3/all.hpp:17:
In file included from /home/sse4/range-v3/include/range/v3/core.hpp:17:
In file included from /home/sse4/range-v3/include/range/v3/begin_end.hpp:21:
In file included from /home/sse4/range-v3/include/range/v3/range_fwd.hpp:20:
/home/sse4/range-v3/include/meta/meta.hpp:1165:22: error: unknown type name 'not_c'
        using not_ = not_c<Bool::type::value>;
                     ^
/home/sse4/range-v3/include/meta/meta.hpp:1165:27: error: expected ';' after alias declaration
        using not_ = not_c<Bool::type::value>;
                          ^
                          ;
/home/sse4/range-v3/include/meta/meta.hpp:1198:22: error: unknown type name 'not_'
        using or_c = not_<std::is_same<integer_sequence<bool, Bools...>,
                     ^
/home/sse4/range-v3/include/meta/meta.hpp:1198:26: error: expected ';' after alias declaration
        using or_c = not_<std::is_same<integer_sequence<bool, Bools...>,
                         ^
                         ;
/home/sse4/range-v3/include/meta/meta.hpp:1206:27: error: unknown type name 'or_c'
        using strict_or = or_c<Bools::type::value...>;
                          ^
/home/sse4/range-v3/include/meta/meta.hpp:1206:31: error: expected ';' after alias declaration
        using strict_or = or_c<Bools::type::value...>;
                              ^
                              ;
/home/sse4/range-v3/include/meta/meta.hpp:1229:32: error: expected a qualified name after 'typename'
            template <typename Bool>
                               ^
/usr/include/X11/Xlib.h:82:14: note: expanded from macro 'Bool'
#define Bool int
             ^
In file included from main.cpp:7:
In file included from /home/sse4/range-v3/include/range/v3/all.hpp:17:
In file included from /home/sse4/range-v3/include/range/v3/core.hpp:17:
In file included from /home/sse4/range-v3/include/range/v3/begin_end.hpp:21:
In file included from /home/sse4/range-v3/include/range/v3/range_fwd.hpp:20:
/home/sse4/range-v3/include/meta/meta.hpp:1229:32: error: expected ',' or '>' in template-parameter-list
/usr/include/X11/Xlib.h:82:14: note: expanded from macro 'Bool'
#define Bool int
             ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```
GCC 5.4.0 has same troubles, but much more long output.

commenting single line `#include <EGL/egl.h>` gives successful build with both GCC & clang.
so the main issue is due to X11/Xlib.h defines:
```
#define Bool int
#define Status int
```

**Comment by CaseyCarter** [comment]
There's nothing unique to range-v3 here: Xlib defines quite a few macros with common names (See https://bugs.freedesktop.org/show_bug.cgi?id=97359) due to incredibly ancient engineering practices. You are likely to have a problem including X11/Xlib.h in the same TU with many headers from libraries that are not specifically intended to be used with Xlib.h.

I suggest that you submit pull request(s) renaming the entities with problematic names.

**Comment by SSE4** [comment]
@CaseyCarter okay, will do. thank you!

**Comment by CaseyCarter** [comment]
Closed by #502.

**Comment by intractabilis** [comment]
`Bool` is not the only conflicting identifier. The following code also conflicts with Xlib.h (from range/v3/view/interface.hpp):
```
        template(bool True = true)(
            requires True AND (Cardinality < 0) AND (Cardinality != infinite) AND
                (!forward_range<D<True>>) AND sized_range<D<True>>)
        constexpr bool empty() //
            noexcept(noexcept(bool(ranges::size(std::declval<D<True> &>()) == 0)))
        {
            return ranges::size(derived()) == 0;
        }
```
This is because Xlib.h defines preprocessor macros `True` and `False`.
